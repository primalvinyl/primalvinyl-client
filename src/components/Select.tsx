import React from 'react';

const Select = ({
    label,
    id,
    value,
    values,
    errors,
    touched,
    required,
    handleChange,
    handleBlur
}: SelectType): React.ReactElement => {
    const elements = values.map(element => {
        return (
            <option
                key={element.value}
                value={element.value}
            >
                {element.displayValue}
            </option>
        );
    });
    return (
        <div className="form-group">
            {label && (<label htmlFor={id}>{label}</label>)}
            <select
                className="form-control"
                name={id}
                id={id}
                value={value}
                onChange={handleChange}
                onBlur={handleBlur}
                required={required}
            >
                {elements}
            </select>
            {errors[id] && touched[id] && (
                <div className="form-field-error">{errors[id]}</div>
            )}
        </div>
    );
};

type SelectType = {
    readonly id: string;
    readonly handleChange: (event: React.ChangeEvent) => void;
    readonly handleBlur: (event: React.FocusEvent) => void;
    readonly values: {value: string; displayValue: string}[];
    readonly label: string;
    readonly value: string;
    readonly required: boolean;
    readonly errors: { [key: string]: string };
    readonly touched: { [key: string]: string };
};

Select.defaultProps = {
    id: '',
    handleChange: (): void => {},
    handleBlur: (): void => {},
    label: '',
    value: '',
    values: [],
    required: false,
    errors: {},
    touched: {}
}

export default Select;
