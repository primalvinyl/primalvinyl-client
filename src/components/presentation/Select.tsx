import React from 'react';

const Select = React.forwardRef<HTMLSelectElement, SelectType>(
    (props, ref): React.ReactElement => {

    const {
        label,
        id,
        value,
        values,
        errors,
        touched,
        required,
        handleChange,
        handleBlur
    } = props;

    let elements: any = [];
    if (values) {
        elements = values.map(element => {
            return (
                <option
                    key={element.value}
                    value={element.value}
                >
                    {element.displayValue}
                </option>
            );
        });
    }

    return (
        <div className="form-group">
            {label && (<label htmlFor={id}>{label}</label>)}
            <select
                className="form-control"
                name={id}
                id={id}
                value={value}
                ref={ref}
                onChange={handleChange}
                onBlur={handleBlur}
                required={required}
            >
                {elements}
            </select>
            {(errors && touched && id) && (errors[id] && touched[id]) && (
                <div className="form-field-error">
                    {errors[id]}
                </div>
            )}
        </div>
    );
});

type SelectType = {
    readonly id?: string;
    readonly values?: {value: string; displayValue: string}[];
    readonly label?: string;
    readonly value?: string;
    readonly required?: boolean;
    readonly errors?: { [key: string]: string };
    readonly touched?: { [key: string]: string };
    readonly handleChange?: (event: React.ChangeEvent<HTMLSelectElement>) => void;
    readonly handleBlur?: (event: React.FocusEvent<HTMLSelectElement>) => void;
};

Select.defaultProps = {
    id: undefined,
    label: undefined,
    value: '',
    values: [],
    required: false,
    errors: {},
    touched: {},
    handleChange: (): void => {},
    handleBlur: (): void => {},
}

export default Select;
