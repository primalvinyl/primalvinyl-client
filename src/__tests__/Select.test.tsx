import * as React from 'react';
import { shallow } from 'enzyme';
import Renderer from 'react-test-renderer';
import Select from '../components/Select';

describe('Select', () => {
    it('renders', () => {
        const myWrapper = shallow(<Select />);
        expect(myWrapper.exists());
    });
    it('matches snapshot', () => {
        const component = Renderer.create(<Select />);
        expect(component.toJSON()).toMatchSnapshot();
    });
});