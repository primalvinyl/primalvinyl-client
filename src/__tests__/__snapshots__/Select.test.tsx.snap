// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select matches snapshot 1`] = `
<div
  className="root"
>
  <div
    className="formControlWrapper"
  >
    <select
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      required={false}
    />
  </div>
</div>
`;
